docs/source/project_documentation/dev_guide/aac_constraints.md

```diff
@@ -0,0 +1,395 @@
+
+# Constraint Definitions
+In order for users to define the self-checking constraints in the AaC DSL, they have to declare constraint rules via the 3 types of constraint definitions. These definitions also provide contextual information for the constraint as well as behaviors and acceptance criteria -- something to leverage for automatically generating functional/integration tests in the future.
+
+## Generation File Safety
+
+We're going to discuss the use of `gen-plugin` below to help you build your constraints.  If you discover you need to update your plugin declaration, you can re-run `gen-plugin` without concern of losing any prior work.  If the output python files already exist, the generator will produce:
+
+  - `.aac_evaluate` files for any user-editable output with an already existing file.  You can look at the changes here and incorporate into your content.
+  - `.aac_backup` files for non-user editable output with an already existing file.  The backup contains the pre-existing content for your review.
+
+Once you're satisfied, run `aac clean <path_to_AaC_plugin_file> --code-output <src_path> --test-output <tests_path>` to remove any `.aac_evaluate` and `.aac_backup` files.
+
+Currently we are not generating documentation for plugins...so you're on your own for that.  Hopefully we'll get some content generation in place in the future to simplify plugin document content management.
+
+## Context Constraints
+Here is an example `context_constraint` definition:
+
+unique_root_keys.aac
+```yaml
+plugin:
+  name: Unique Root Keys
+  package: aac.plugins.unique_root_keys
+  description: An AaC plugin that enables validating all required fields are present in a definition.
+  context_constraints:
+    - name: Root key names are unique
+      description: Check every definition to ensure there are no duplicate root keys defined in the AaC language.
+```
+
+To implement the context constraint, run `aac gen-plugin <path-to-unique_root_keys.aac> --code-output <src-path> --test-output <tests-path>`.  Note that the package declaration will be used to establish sub-directories under your `--code-output` and `--test-output` paths.  By default, the `gen-plugin` command will prompt you to confirm the target output paths prior to attempting to generate any files.
+
+The `gen-plugin` command will auto-generate all the AaC boilerplate code to integrate your new command into the AaC plugin framework and will provide you an implementation stub to populate.  It will also generate a unit test and feature files based on your declaration.  This particular example does not demonstrate the use of feature files.
+
+Generated implementation stub:
+```python
+"""The AaC Unique Root Keys plugin implementation module."""
+# NOTE: It is safe to edit this file.
+# This file is only initially generated by aac gen-plugin, and it won't be overwritten if the file already exists.
```

Comment:
I disagree...and it is accurate.  The challenge is if a user changes their model after having run gen-plugin, the interface to their implementation may change as a result.  I spent a lot of time (a lot) trying to figure out how to generate new implementation boilerplate, diff, and merge it into the user code without botching their logic...and was not successful.  I settled on the "backup" and "eval" strategy so that we will never blast away anyone's hard work.  It does leave the user to actually look at the "eval" file and manually merge changes...but there's no other feasible solution that I could find.  Open to suggestions on a better strategy.